### ISSUE TITLE
Investigate dynamic generation of test X509 certificates in Bridge

### ISSUE THREAD

**Issue by iamjasonp** [dotnet/wcf#324]
Our current test X509 certificates are generated using `makecert.exe` and committed to the repo. This poses a few problems: 

1. Certs expire after a set period of time (currently, two months after generation of the certificate) 
2. Certs are tied to localhost only and we do not generate certs for the fully qualified name of a machine
3. Certificates are not generated with a Subject Alternative Name (#320) or Certificate Revocation List (#310) reference

We want to investigate a solution for this issue, and to also allow the Bridge to supply certificates to a caller (script, perhaps?) when asked for them. 

To investigate
---

Methods to generate cert: 
1. marshal to Windows Crypto APIs 
2. Bouncy Castle
3. OpenSSL

How to interrogate the Bridge for certs: 
> e.g., as soon as the Bridge knows the BridgeResourceFolder via the /config enpoint, it could just generate the certs, and the scripts could pull them back out.

The assumption is that there is no need for Bridge.exe to be cross-plat at the moment.

**Comment by iamjasonp** [comment]
Currently blocked on legal approval as we want to depend on an external library for cert generation

**Comment by ericlaw1979** [comment]
Kinda jumping in out of the blue here, but if you're using Win7+ you could take the approach undertaken by Fiddler and use the CertEnroll library that ships with Windows. Wrapping it from C# is pretty trivial and it allows you a high degree of customization of the fields of the certs.

For non-Windows platforms, Fiddler uses BouncyCastle, although I ended up stripping out 75% of the library unrelated to x509.

**Comment by iamjasonp** [comment]
@ericlaw1979 - thanks! I should have been better about updating this issue as time has gone on, as significant work has ended up going into this and we're ready for go-live save for approval from Microsoft LCA. 

We decided to focus our investigation on the two technologies that we know we could integrate into the code easily. Here's what we evaluated: 

*CertEnroll*
* :heavy_plus_sign: Shipped as part of Windows
* :heavy_plus_sign: If there are bugs in the crypto libraries, no need to update them ourselves as they'll be part of Windows updates 
* :heavy_minus_sign: No way to generate Certificate Revocation Lists 
* :heavy_minus_sign: Seems to automatically litter the Intermediate Root Authorities store with certificates if a root authority is generated 
* :heavy_minus_sign: Certificates generated always go into the cert store and we need to export it and then delete it (not all certs need to live on the generating machine - since we wanted to centralise cert generation)
* :heavy_minus_sign: Code in C# is a bit more complex and harder to read

*Bouncy Castle:*
* :heavy_plus_sign: Good API to generate certificates 
* :heavy_plus_sign: Built-in way to generate certificate revocation lists
* :heavy_minus_sign: Last binary release was in 2011 (pro since it hasn't needed to change much?) 
* :heavy_minus_sign: Need additional MS approvals to use 3rd party libraries in the product
* I haven't tried to incorporate the source code directly. 

*OpenSSL*
*Not evaluated - documentation online seems to allow for use only by command line*

**Design** 
We could either generate certificates on the client or the server (Bridge) side. Each had its own advantages, but after a lot of thought, a decision was taken to model the certificate generation on the Bridge
* We could model certificate generation this as a Bridge IResource 
* Generate certificates that were signed by a single authority throughout the life of one bridge instance, without having to share the private key of the root cert via file share or http
* Prevent the generation of arbitrary certificates by parties not participating in the testing (the Bridge limits access to the /resource endpoint)
* Limit the lifetime of each certificate/root authority to prevent security issues on dev boxes
* The use of the Bridge IResource allows us to expand capabilities of each of the endpoints as we need to

**Resources**
`CertificateAuthorityResource`
`PUT` create the certificate generator and certificate authority. Returns the thumbprint
`GET` retrieve the certificate from the server, base64 encoded as part of a json key-value payload

`MachineCertificateResource`
`PUT` creates a certificate for a set of comma separated list of `subjects`. Returns the thumbprint
`GET` takes a `subject`, retrieves the certificate, base64 encoded as part of a json key-value payload

`CertificateRevocationListResource`
`PUT` Creates the CRL
`GET` Returns the CRL as raw data (application/octet-stream)

**Decision** 
We want the ability to test for certificate revocation lists - WCF exposes APIs `X509ServiceCertificateAuthentication.CertificateValidationMode` and `X509ServiceCertificateAuthentication.RevocationMode` to control how WCF does certificate validation and revocation checking. This ultimately swayed our decision to move to using Bouncy Castle

**Comment by iamjasonp** [comment]
@ericlaw1979 - btw, great fan of Fiddler :smile: I use Fiddler tool a _lot_, pretty much for as long as I remember working on anything with HTTP in network communications.

**Comment by ericlaw1979** [comment]
Thanks for the detail! For what its worth, I hit one problem in BouncyCastle I had to fix-- the default code didn't work on Thai systems: http://www.bouncycastle.org/jira/browse/BMA-113?filter=-2

From my changelist:
// v2.0.3 // Fixed Thai date formatting problem

// v2.0.2 // Update padding rules per the latest BouncyCastle CVS snapshot (1.7.1) (DotNetUtilities File v1.16)

// v2.0.1.1 // Ensure that PrivateKey's byte array lengths are compatible with Windows/.NET import (see ToRSAParameters).

I also had to take some care when setting the certificate serial number - BC accepts serial numbers that browsers reject (IIRC, small numbers and negative numbers both cause issues).

**Comment by iamjasonp** [comment]
#394 is the PR that's going to implement this capability for us - please feel free to chime in :smile: though it's a large PR. All the cert generation code is encapsulated in the CertificateGenerator.cs file 

This needs to go in first to unblock some of our test scenarios, so hopefully in the meantime nobody runs into these issues you mentioned... 

The dependency we're taking is on this https://www.nuget.org/packages/BouncyCastle/ since it seems like it's the most authoritative-looking Nuget package available to us - it's at 1.7.0. We can't just take the latest since we have to have this package restoreable instead of checking in a DLL. 

For the longer term, I'm almost certain we'll have to take in the fixes you mentioned - we'll probably take these as a separate issue. I will probably file a bug over in BC to have them update the Nuget packages so we can depend on them. 

... interestingly, I did stumble on an issue in the BC repo where someone has suggested [BC be merged into CoreFx](http://www.bouncycastle.org/jira/browse/BMA-131). Really interesting thought, that one, and I'll bring that up as a community suggestion to CoreFx
